File: src/test/java/com/github/rizvn/pgcache/PgCacheTest.java
package com.github.rizvn.pgcache;

import com.github.rizvn.DbTest;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;
import org.junit.jupiter.api.DisplayName;
import org.slf4j.Logger;
import org.slf4j.LoggerFactory;

import static org.junit.jupiter.api.Assertions.*;

@DisplayName("PgCache Tests")
class PgCacheTest extends DbTest {

    private static final Logger logger = LoggerFactory.getLogger(PgCacheTest.class);
    private PgCache pgCache;

    @BeforeEach
    void setUp() {
        // Create PgCache instance with the test dataSource from DbTest
        pgCache = new PgCache(dataSource, "c_test", 600); // 10 minutes TTL
    }

    @Test
    @DisplayName("Put Value")
    void testPutValue() {
        // Arrange
        String testId = "test_key";
        byte[] testContent = "test_value".getBytes();

        // Act
        pgCache.put(testId, testContent);

        // Assert - no exception means success
        logger.info("Successfully put value for key: {}", testId);
    }

    @Test
    @DisplayName("Get Value")
    void testGetValue() {
        // Arrange
        String testId = "test_key";
        byte[] testContent = "test_value".getBytes();
        pgCache.put(testId, testContent);

        // Act
        byte[] content = pgCache.get(testId);

        // Assert
        assertNotNull(content, "Expected to find key " + testId + ", but it was not found");
        byte[] expectedContent = "test_value".getBytes();
        assertArrayEquals(expectedContent, content,
            "Expected content " + new String(expectedContent) + ", got " + new String(content));
    }

    @Test
    @DisplayName("PutWithTTL Value")
    void testPutWithTTL() throws InterruptedException {
        // Arrange
        String testId = "test_key_ttl";
        byte[] testContent = "ttl_value".getBytes();

        // Act - put with 1 second TTL
        pgCache.putWithTtl(testId, testContent, 1);

        // Assert - value should be found immediately
        byte[] content = pgCache.get(testId);
        assertNotNull(content, "Expected to find key " + testId + " with correct value");
        assertEquals("ttl_value", new String(content), "Expected correct value");

        // Wait for TTL to expire
        logger.info("Waiting for 2 seconds for TTL to expire");
        Thread.sleep(2000);

        // Assert - value should be expired
        content = pgCache.get(testId);
        assertNull(content, "Expected key " + testId + " to be expired");
    }

    @Test
    @DisplayName("Delete Value")
    void testDeleteValue() {
        // Arrange
        String testId = "delete_key";
        byte[] testContent = "delete_value".getBytes();
        pgCache.put(testId, testContent);

        // Act
        pgCache.delete(testId);

        // Assert
        byte[] content = pgCache.get(testId);
        assertNull(content, "Expected key " + testId + " to be deleted");
    }

    @Test
    @DisplayName("DeleteExpired")
    void testDeleteExpired() throws InterruptedException {
        // Arrange
        String testId = "expired_key";
        byte[] testContent = "expired_value".getBytes();
        pgCache.putWithTtl(testId, testContent, 1);

        // Wait for expiration
        logger.info("Waiting for 2 seconds for TTL to expire");
        Thread.sleep(2000);

        // Act
        pgCache.deleteExpired();

        // Assert - value should not be found (even without expiration check)
        byte[] content = pgCache.get(testId);
        assertNull(content, "Expected expired key " + testId + " to be deleted");
    }

    @Test
    @DisplayName("ClearCacheStore")
    void testClearCacheStore() {
        // Arrange
        pgCache.put("clear_key", "clear_value".getBytes());

        // Act
        pgCache.clearCacheStore();

        // Assert
        byte[] content = pgCache.get("clear_key");
        assertNull(content, "Expected cache store to be cleared");
    }

    @Test
    @DisplayName("DropCacheStore")
    void testDropCacheStore() {
        // Arrange
        pgCache.put("drop_key", "drop_value".getBytes());

        // Act
        pgCache.dropCacheStore();

        // Assert - table should not exist, so put should throw exception
        assertThrows(RuntimeException.class,
            () -> pgCache.put("drop_key", "drop_value".getBytes()),
            "Expected RuntimeException after dropping cache table");
    }
}

File: pgcache/PgCache_test.go
package pgcache_test

import (
	"context"
	"database/sql"
	"testing"
	"time"

	"github.com/rizvn/pgutils/pgcache"
	"github.com/rizvn/pgutils/testutil"
)

func TestPgCache(t *testing.T) {
	// Start Postgres test container
	ctr, dsn := testutil.StartPgTestContainer()
	defer func() { _ = ctr.Terminate(context.Background()) }()

	dbPool, err := sql.Open("pgx", dsn)
	if err != nil {
		t.Fatal("failed to create pgx pool")
	}
	dbPool.SetMaxOpenConns(20)

	// Create PgCache instance
	// 600 seconds TTL (10 minutes) for testing
	pgCache := pgcache.NewPgCache(dbPool, "c_test", pgcache.WithTTL(600))

	err = pgCache.CreateCacheTable()
	if err != nil {
		t.Fatalf("failed to create cache table: %v", err)
	}

	t.Run("Put Value", func(t *testing.T) {
		// Test Put
		testID := "test_key"
		testContent := []byte("test_value")
		pgCache.Put(testID, testContent) // 5 minutes TTL
	})

	t.Run("Get Value", func(t *testing.T) {
		// Test Get
		testID := "test_key"
		content, found := pgCache.Get(testID)
		if !found {
			t.Errorf("Expected to find key %s, but it was not found", testID)
		}
		expectedContent := []byte("test_value")
		if string(content) != string(expectedContent) {
			t.Errorf("Expected content %s, got %s", expectedContent, content)
		}
	})

	t.Run("PutWitTTL Value", func(t *testing.T) {
		testID := "test_key_ttl"
		testContent := []byte("ttl_value")
		err = pgCache.PutWitTTL(testID, testContent, 1) // 1 second TTL

		if err != nil {
			t.Fatalf("failed to put value with TTL: %v", err)
		}

		content, found := pgCache.Get(testID)
		if !found || string(content) != "ttl_value" {
			t.Errorf("Expected to find key %s with correct value", testID)
		}
		time.Sleep(2 * time.Second)
		_, found = pgCache.Get(testID)
		if found {
			t.Errorf("Expected key %s to be expired", testID)
		}
	})

	t.Run("Delete Value", func(t *testing.T) {
		testID := "delete_key"
		testContent := []byte("delete_value")
		pgCache.Put(testID, testContent)
		err = pgCache.Delete(testID)
		if err != nil {
			t.Fatalf("failed to delete key %s: %v", testID, err)
		}
		_, found := pgCache.Get(testID)
		if found {
			t.Errorf("Expected key %s to be deleted", testID)
		}
	})

	t.Run("DeleteExpired", func(t *testing.T) {
		testID := "expired_key"
		testContent := []byte("expired_value")
		err = pgCache.PutWitTTL(testID, testContent, 1)
		if err != nil {
			t.Fatalf("failed to put value with TTL: %v", err)
		}
		time.Sleep(5 * time.Second)
		err = pgCache.DeleteExpired()
		if err != nil {
			t.Fatalf("failed to delete expired keys: %v", err)
		}
		_, found := pgCache.Get(testID)
		if found {
			t.Errorf("Expected expired key %s to be deleted", testID)
		}
	})

	t.Run("ClearCacheStore", func(t *testing.T) {
		pgCache.Put("clear_key", []byte("clear_value"))
		err = pgCache.ClearCacheStore()
		if err != nil {
			t.Fatalf("failed to clear cache store: %v", err)
		}
		_, found := pgCache.Get("clear_key")
		if found {
			t.Errorf("Expected cache store to be cleared")
		}
	})

	t.Run("DropCacheStore", func(t *testing.T) {
		pgCache.Put("drop_key", []byte("drop_value"))
		err = pgCache.DropCacheStore()

		if err != nil {
			t.Fatalf("failed to drop cache store: %v", err)
		}

		err = pgCache.Put("drop_key", []byte("drop_value"))
		if err == nil {
			t.Errorf("Expected error when putting value after dropping cache store, but got none")
		}
	})

}
